src/OpenTelemetry.AutoInstrumentation/Loading/MySqlDataInitializer.cs

```diff
@@ -37,5 +37,26 @@ public override void Initialize(ILifespanManager lifespanManager)
 
         lifespanManager.Track(instrumentation);
     }
+
+    private class MySqlDataRequiredAssemblies : IRequiredAssemblies
```

Comment:
1. From the reader's perspective is less safe. Are we sure that it is not possible to load two times the same assembly in the same domain  (e.g. if one loads `"MySql.Data"` via reflection)? If so then we do not even need any state -> we just load the instrumentation when `assemblyName == "MySql.Data"` and then remove the handler. BTW both implementation using `lock` and `Interlocked.Increment` were protected from such a scenario.
1. Do we expect that we will require more assemblies than one?

Maybe we could simply change `public IRequiredAssemblies RequiredAssemblies` to `public string RequiredAssemblyName` in`InstrumentationInitializer`. This would allow us the get rid of `MySqlDataRequiredAssemblies` and `IRequiredAssemblies`. It should reduce the complexity and improve performance a little (interfaces bring a _little_ overhead and we would have fewer objects).

Reply:
> 1. Are we sure that it is not possible to load two times the same assembly in the same domain  (e.g. if one loads `"MySql.Data"` via reflection)?

According to https://docs.microsoft.com/en-us/dotnet/framework/deployment/best-practices-for-assembly-loading it is possible to load an assembly more than once. I think those scenarios are edge cases, but they are possible.

> 2\. Do we expect that we will require more assemblies than one?

I don't know if we have a need for that yet. If we do not have a need for that yet, then we should probably use one of the simpler approaches that you recommended.